lib/api/api_auth.go

```diff
@@ -61,61 +116,94 @@ func basicAuthAndSessionMiddleware(cookieName string, guiCfg config.GUIConfigura
 			}
 		}
 
-		l.Debugln("Sessionless HTTP request with authentication; this is expensive.")
-
-		error := func() {
-			time.Sleep(time.Duration(rand.Intn(100)+100) * time.Millisecond)
-			w.Header().Set("WWW-Authenticate", "Basic realm=\"Authorization Required\"")
-			http.Error(w, "Not Authorized", http.StatusUnauthorized)
+		// Fall back to Basic auth if provided
+		if username, ok := attemptBasicAuth(r, guiCfg, ldapCfg, evLogger); ok {
+			createSession(cookieName, username, guiCfg, evLogger, w, r)
+			next.ServeHTTP(w, r)
+			return
 		}
 
-		username, password, ok := r.BasicAuth()
-		if !ok {
-			error()
+		// Some browsers don't send the Authorization request header unless prompted by a 401 response.
+		// This enables https://user:pass@localhost style URLs to keep working.
+		if guiCfg.SendBasicAuthPrompt {
+			unauthorized(w)
```

Comment:
Hmyeah, this seems to be a side effect of that before, every HTTP request was behind basic auth, so when you attempt to load `/index.html`, the browser sees the 401 response and shows the prompt, then it gives up after a number of tries. Now it's instead an AJAX request that triggers the prompt, which the browser seems to handle differently. In both Firefox and Chrome you can cancel out of it, though.

If we want to keep the old behaviour, that should be fairly easy to do. In that case I would propose renaming the `SendBasicAuthPrompt` setting to `UseBasicAuth` and have it override the HTML form so that again everything is behind basic auth (meaning the HTML login form will be unavailable).

Reply:
I'd prefer if both the HTML form and basic auth were available if configured, not as an alternative disabling the HTML form.  So what happens after hitting "cancel" now?

I'm okay with getting the basic auth prompt repeatedly, without limit.  Just wanted to clarify why there's a difference.  Thanks for the explanation.

Is it possible to send a 401 without another basic auth request?  If so, the backend could implement the failure counter, if we have an indication that it's the same session failing multiple times?

Reply:
> I'd prefer if both the HTML form and basic auth were available if configured, not as an alternative disabling the HTML form. So what happens after hitting "cancel" now?

Agreed, and that's how it works as currently implemented if the `sendBasicAuthPrompt` setting is set to `true`. If you press "cancel" in the auth dialog, it just closes until you reload the page (or presumably until another AJAX request is sent, but the current initialization JS bails out early if that first request fails).

> Is it possible to send a 401 without another basic auth request?

Well, it is technically within our power to do that, but it's against the [HTTP spec](https://www.rfc-editor.org/rfc/rfc9110#status.401):

>The server generating a 401 response MUST send a [WWW-Authenticate](https://www.rfc-editor.org/rfc/rfc9110#field.www-authenticate) header field ([Section 11.6.1](https://www.rfc-editor.org/rfc/rfc9110#field.www-authenticate)) containing at least one challenge applicable to the target resource.

We could switch to 403 after a few failed attempts, but I don't really see the value in attempting to emulate the old behaviour.

Reply:
In my case, it show two basic auth requests if cancelled, then goes to the login form.  Same happens when entering empty credentials.  Wrong combinations can be tried endlessly, there is no abort after a certain number of retries.  Tested on Firefox.

So I guess this is good to go.